ddclient.in

```diff
@@ -4158,154 +4164,208 @@ EoEXAMPLE
-        );
-        my $ipv4 = delete $config{$h}{'wantipv4'};
-        my $ipv6 = delete $config{$h}{'wantipv6'};
+    # Local subroutine to set the name and check the zone
+    my $set_name_and_check_zone = sub {
+        my ($hostname, $config) = @_;
+
+        # Create a copy of the $config hash
+        my %config_copy = %$config;
+
+        # If $config_copy{zone} is empty (''), return $name = ''
+        if ($check_zone->(\%config_copy)){
+            debug("Zone not defined or empty for hostname: %s", $hostname);
+            return ('', \%config_copy);
+        }
+
+        # If $config_copy{zone} is set, create $name by trimming the zone from the hostname.
+        my $name = $hostname;
+        my $zone = $config_copy{'zone'};
+        if ($hostname =~ /\Q$zone\E$/) {
+            $name =~ s/\.?\Q$zone\E$//;
+        } else {
+            failed("Hostname %s does not end with the zone %s, cannot be trimmed", $hostname, $zone);
+        }
 
+        return ($name, \%config_copy);
+    };
+
+    ## Local subroutine to rewrite the domain name if zone is defined
+    my $set_domain = sub {
+        my ($hostname, $config) = @_;
+
+        if ($check_zone->($config)){
+            debug("Zone not defined or empty for hostname: %s", $hostname);
+            debug("Setting zone to hostname: %s", $hostname);
+            $config->{'zone'} = $hostname;
+        }
+        return ($config);
+    };
+    ## Update each configured host (hosts cannot be grouped across domains on this API)
+    foreach my $hostname (@_) {
+        my $config = $config{$hostname}; # isolate the config for this hostname
+        my $ipv4 = delete $config->{'wantipv4'};
+        my $ipv6 = delete $config->{'wantipv6'};
+       
         # Updates for ipv4 and ipv6 need to be combined in a single API call, create Hash of Arrays for tracking
         my %total_payload;
 
-        foreach my $ip ($ipv4, $ipv6){
-            next if (!$ip);
-            my $ipv  = ($ip eq ($ipv6 // '')) ? '6' : '4';
-            my $type = ($ip eq ($ipv6 // '')) ? 'AAAA' : 'A';
+        foreach my $ip_address ($ipv4, $ipv6){
+            next if (!$ip_address);
+            my $ipv  = ($ip_address eq ($ipv6 // '')) ? '6' : '4';
+            my $type = ($ip_address eq ($ipv6 // '')) ? 'AAAA' : 'A';
+
+            info("Going to update IPv$ipv address to %s for %s.", $ip_address, $hostname);
+            $config->{"status-ipv$ipv"} = 'failed';
 
-            info("Going to update IPv$ipv address to %s for %s.", $ip, $h);
-            $config{$h}{'status-ipv$ipv'} = 'failed';
-            $total_payload{$ipv} = {
-                name => $h,
-                type => $type,
-                content => $ip,
-                ttl => $config{$h}{'ttl'},
-            };
+            # One key per ipv (4 or 6)
+            my ($name, $modified_config) = $set_name_and_check_zone->($hostname, $config);
```

Comment:
The helper functions and config copy seem unnecessary to me.  Wouldn't something more direct like the following work?

```perl
            my $zone = $config->{'zone'} || $hostname;
            my $name = $hostname;
            if ($name !~ s/(?:^|\.)\Q$zone\E$//) {
                failed("Hostname %s does not end with the zone %s, cannot be trimmed", $hostname, $zone);
                next;
            }
```

Reply:
Yes, I've made this shorter / more to the point.